File: internal/engine/domain/lexer.go
// Package domain implements the domain expression language shared across
// GoERP: ABAC policy conditions, host.orm.search domain arguments, and
// (in other contexts, not this package) shell-interpreted view/field
// conditions. See manifest-spec.md §8 for the grammar.
package domain

import (
	"fmt"
	"strings"
	"unicode"
)

type tokenKind int

const (
	tokEOF tokenKind = iota
	tokIdent
	tokString
	tokNumber
	tokDot
	tokLParen
	tokRParen
	tokComma
	tokEq
	tokNeq
	tokLt
	tokGt
	tokLte
	tokGte

	// keywords
	tokAnd
	tokOr
	tokNot
	tokIn
	tokIs
	tokNull
	tokTrue
	tokFalse
	tokLike
	tokIlike
	tokChildOf
	tokParentOf
)

var keywords = map[string]tokenKind{
	"and":       tokAnd,
	"or":        tokOr,
	"not":       tokNot,
	"in":        tokIn,
	"is":        tokIs,
	"null":      tokNull,
	"true":      tokTrue,
	"false":     tokFalse,
	"like":      tokLike,
	"ilike":     tokIlike,
	"child_of":  tokChildOf,
	"parent_of": tokParentOf,
}

type token struct {
	kind tokenKind
	text string
	pos  int
}

type lexer struct {
	src string
	pos int
}

func newLexer(src string) *lexer {
	return &lexer{src: src}
}

func (l *lexer) tokenize() ([]token, error) {
	var toks []token
	for {
		tok, err := l.next()
		if err != nil {
			return nil, err
		}
		toks = append(toks, tok)
		if tok.kind == tokEOF {
			return toks, nil
		}
	}
}

func (l *lexer) next() (token, error) {
	l.skipSpace()
	if l.pos >= len(l.src) {
		return token{kind: tokEOF, pos: l.pos}, nil
	}

	start := l.pos
	c := l.src[l.pos]

	switch {
	case c == '\'':
		return l.lexString()
	case c == '.':
		l.pos++
		return token{kind: tokDot, text: ".", pos: start}, nil
	case c == '(':
		l.pos++
		return token{kind: tokLParen, text: "(", pos: start}, nil
	case c == ')':
		l.pos++
		return token{kind: tokRParen, text: ")", pos: start}, nil
	case c == ',':
		l.pos++
		return token{kind: tokComma, text: ",", pos: start}, nil
	case c == '=':
		l.pos++
		return token{kind: tokEq, text: "=", pos: start}, nil
	case c == '!':
		if l.peekAt(1) == '=' {
			l.pos += 2
			return token{kind: tokNeq, text: "!=", pos: start}, nil
		}
		return token{}, fmt.Errorf("domain: unexpected character %q at position %d", c, start)
	case c == '<':
		if l.peekAt(1) == '=' {
			l.pos += 2
			return token{kind: tokLte, text: "<=", pos: start}, nil
		}
		l.pos++
		return token{kind: tokLt, text: "<", pos: start}, nil
	case c == '>':
		if l.peekAt(1) == '=' {
			l.pos += 2
			return token{kind: tokGte, text: ">=", pos: start}, nil
		}
		l.pos++
		return token{kind: tokGt, text: ">", pos: start}, nil
	case isDigit(c):
		return l.lexNumber()
	case isIdentStart(c):
		return l.lexIdent()
	default:
		return token{}, fmt.Errorf("domain: unexpected character %q at position %d", c, start)
	}
}

func (l *lexer) lexString() (token, error) {
	start := l.pos
	l.pos++ // consume opening quote
	var sb strings.Builder
	for {
		if l.pos >= len(l.src) {
			return token{}, fmt.Errorf("domain: unterminated string literal starting at position %d", start)
		}
		c := l.src[l.pos]
		if c == '\'' {
			// SQL-style escaping: '' inside a string is a literal quote.
			if l.peekAt(1) == '\'' {
				sb.WriteByte('\'')
				l.pos += 2
				continue
			}
			l.pos++
			return token{kind: tokString, text: sb.String(), pos: start}, nil
		}
		sb.WriteByte(c)
		l.pos++
	}
}

func (l *lexer) lexNumber() (token, error) {
	start := l.pos
	for l.pos < len(l.src) && (isDigit(l.src[l.pos]) || l.src[l.pos] == '.') {
		l.pos++
	}
	return token{kind: tokNumber, text: l.src[start:l.pos], pos: start}, nil
}

func (l *lexer) lexIdent() (token, error) {
	start := l.pos
	for l.pos < len(l.src) && isIdentPart(l.src[l.pos]) {
		l.pos++
	}
	text := l.src[start:l.pos]
	if kind, ok := keywords[strings.ToLower(text)]; ok {
		return token{kind: kind, text: text, pos: start}, nil
	}
	return token{kind: tokIdent, text: text, pos: start}, nil
}

func (l *lexer) skipSpace() {
	for l.pos < len(l.src) && unicode.IsSpace(rune(l.src[l.pos])) {
		l.pos++
	}
}

func (l *lexer) peekAt(offset int) byte {
	if l.pos+offset >= len(l.src) {
		return 0
	}
	return l.src[l.pos+offset]
}

func isDigit(c byte) bool { return c >= '0' && c <= '9' }

func isIdentStart(c byte) bool {
	return c == '_' || (c >= 'a' && c <= 'z') || (c >= 'A' && c <= 'Z')
}

func isIdentPart(c byte) bool {
	return isIdentStart(c) || isDigit(c)
}

File: shell/packages/sdk/src/domain/lexer.ts
import { DomainExpressionError } from "./errors.js";

export type TokenKind =
  | "eof"
  | "ident"
  | "string"
  | "number"
  | "dot"
  | "lparen"
  | "rparen"
  | "comma"
  | "eq"
  | "neq"
  | "lt"
  | "gt"
  | "lte"
  | "gte"
  | "plus"
  | "minus"
  | "star"
  | "slash"
  | "and"
  | "or"
  | "not"
  | "in"
  | "is"
  | "null"
  | "true"
  | "false"
  | "like"
  | "ilike"
  | "child_of"
  | "parent_of";

export interface Token {
  kind: TokenKind;
  text: string;
  pos: number;
}

// Keywords are case-insensitive, identifiers are not, as in the Go lexer.
const KEYWORDS: ReadonlyMap<string, TokenKind> = new Map<string, TokenKind>([
  ["and", "and"],
  ["or", "or"],
  ["not", "not"],
  ["in", "in"],
  ["is", "is"],
  ["null", "null"],
  ["true", "true"],
  ["false", "false"],
  ["like", "like"],
  ["ilike", "ilike"],
  ["child_of", "child_of"],
  ["parent_of", "parent_of"],
]);

const NUMBER_LITERAL = /^\d+(\.\d+)?$/;

function isDigit(c: string): boolean {
  return c >= "0" && c <= "9";
}

function isIdentStart(c: string): boolean {
  return c === "_" || (c >= "a" && c <= "z") || (c >= "A" && c <= "Z");
}

function isIdentPart(c: string): boolean {
  return isIdentStart(c) || isDigit(c);
}

function isSpace(c: string): boolean {
  return c === " " || c === "\t" || c === "\n" || c === "\r" || c === "\v" || c === "\f";
}

const SINGLE_CHAR: Record<string, TokenKind> = {
  ".": "dot",
  "(": "lparen",
  ")": "rparen",
  ",": "comma",
  "=": "eq",
};

const ARITHMETIC_CHAR: Record<string, TokenKind> = {
  "+": "plus",
  "-": "minus",
  "*": "star",
  "/": "slash",
};

export interface LexOptions {
  // `+ - * /` exist only in the computed_display grammar (manifest-spec.md §8).
  arithmetic: boolean;
}

export function tokenize(src: string, options: LexOptions): Token[] {
  const tokens: Token[] = [];
  let pos = 0;

  const fail = (message: string, at: number): never => {
    throw new DomainExpressionError("parse", message, src, at);
  };

  while (true) {
    while (pos < src.length && isSpace(src.charAt(pos))) pos++;
    if (pos >= src.length) {
      tokens.push({ kind: "eof", text: "", pos });
      return tokens;
    }

    const start = pos;
    const c = src.charAt(pos);
    const next = src.charAt(pos + 1);

    const single = SINGLE_CHAR[c];
    if (single !== undefined) {
      pos++;
      tokens.push({ kind: single, text: c, pos: start });
      continue;
    }

    const arithmetic = options.arithmetic ? ARITHMETIC_CHAR[c] : undefined;
    if (arithmetic !== undefined) {
      pos++;
      tokens.push({ kind: arithmetic, text: c, pos: start });
      continue;
    }

    if (c === "!") {
      if (next !== "=") fail(`unexpected character "!"`, start);
      pos += 2;
      tokens.push({ kind: "neq", text: "!=", pos: start });
      continue;
    }

    if (c === "<" || c === ">") {
      const orEqual = next === "=";
      pos += orEqual ? 2 : 1;
      const kind: TokenKind = c === "<" ? (orEqual ? "lte" : "lt") : orEqual ? "gte" : "gt";
      tokens.push({ kind, text: orEqual ? `${c}=` : c, pos: start });
      continue;
    }

    if (c === "'") {
      pos++;
      let value = "";
      while (true) {
        if (pos >= src.length) fail("unterminated string literal", start);
        const ch = src.charAt(pos);
        if (ch === "'") {
          // SQL-style escaping: '' inside a string is a literal quote.
          if (src.charAt(pos + 1) === "'") {
            value += "'";
            pos += 2;
            continue;
          }
          pos++;
          break;
        }
        value += ch;
        pos++;
      }
      tokens.push({ kind: "string", text: value, pos: start });
      continue;
    }

    if (isDigit(c)) {
      while (pos < src.length && (isDigit(src.charAt(pos)) || src.charAt(pos) === ".")) pos++;
      const text = src.slice(start, pos);
      if (!NUMBER_LITERAL.test(text)) fail(`invalid number literal "${text}"`, start);
      tokens.push({ kind: "number", text, pos: start });
      continue;
    }

    if (isIdentStart(c)) {
      while (pos < src.length && isIdentPart(src.charAt(pos))) pos++;
      const text = src.slice(start, pos);
      tokens.push({ kind: KEYWORDS.get(text.toLowerCase()) ?? "ident", text, pos: start });
      continue;
    }

    fail(`unexpected character "${c}"`, start);
  }
}
